halibot: introduce CommandModule subclass of HalModule

richteer opened the pull request:
Initial implementation of the CommandModule class of HalModule.

To implement a CommandModule:
- Inherit from the halibot.CommandModule class
- Define a class variable that is a
  dict: "command" -> func(self, string, msg=msg)

Example:

``` python
class Foo(CommandModule):
    def init(self):
        self.commands = {"hello": self.hello_}

    def hello_(self, string, msg=None):
        return self.reply(msg, body="Hello World!")
```

Note: it is recommended practice to trail command methods with an underscore
so as not to accidentally override internal halobject methods.

richteer commented:
Further TODO is to support more complex commands. Example, a `!command` may have subcommands like `!command add <data>` or `!command subcommand add <data>`.

So, more complex structuring might be needed, will think about more in the future.

richteer commented:
Command table could be more robust. Maybe make the value a tuple of (func, "Description"), so a help module could poll this information.

sjrct commented:
Should the command dictionary exist on the class rather than the instance?

EDIT from richteer since I can't reply to this anymore...: Yes, the command dictionary should be on instance, not the class. A module should have the ability to remove commands at runtime and not affect other instances.

sjrct commented on halibot/__init__.py:
```diff
@@ -2,3 +2,4 @@
 from .halagent import HalAgent
 from .halmodule import HalModule
 from .message import Context, Message
+from .commandmodule import CommandModule, AsArgs
```
We shouldn't need this import, right?

sjrct commented on halibot/commandmodule.py:
```diff
@@ -0,0 +1,59 @@
+# Decorator for converting the string stripped of the command to an argument list
+def AsArgs(func):
+	def wrapper(self, string, msg=None):
+		args = string.split(" ")
```
This can give us empty arguments. Could filter out empties after the split, or strip blanks and split over the regex `" +"`.

Something like this: `args = re.split(" +", string.strip())`
Or like this: `args = [x if x != '' for x in string.split(" ")]`
Or or like this: `args = filter(lambda x: x != '', string.split(" "))`

None of those are tested, by the way.

richteer replied:
Still thinking on this one. I agree that a regex would be better than `str.split()` in all cases in here. Only thing I am hung up on is, where to stick the compiled regex. There's no point in compiling it on every receive, so it should probably be a singleton somewhere?

sjrct replied:
I don't think it really matters... you can just stick the regex in the top-level of the file.

sjrct commented on halibot/commandmodule.py:
```diff
@@ -0,0 +1,59 @@
+		super().__init__(hal, conf=conf)
+
+		# Clear the static var to an empty table
+
```
That's a lot of code to just clear the static var to an empty table.

sjrct commented:
Looking at your code in the example it seems like the `commands` dict should be attached to the Module in the example, as it is attached to the class in `CommandModule`.

sjrct commented on halibot/commandmodule.py:
```diff
@@ -0,0 +1,59 @@
+			body.append("")
+
+		# This maybe should prevent a empty string as a key...
+		func = self.commands.get(body[0][1:])
```
I don't think it is necessary to check for an empty string. If somebody puts `"": bar` in their command table then it should run `bar` when they type something like `! quaz quax`. Although that is probably not a good idea to do in practice.

richteer replied:
I don't like the idea of allowing empty commands via this. I see CommandModule as a way to simplify creating modules that use the `!foo bar` -> `.foo(bar)` pattern, and thus an empty command string should be a error that is caught.

Nothing is stopping someone from using an empty command, just probably not with this class.

sjrct replied:
Maybe `! foo` == `!foo`?

sjrct commented on halibot/commandmodule.py:
```diff
@@ -0,0 +1,59 @@
+		func = self.commands.get(body[0][1:])
+
+		if func:
+			func(body[1], msg=msg)
```
`body.get(1, "")`? And get rid of lines 45-47?

richteer replied:
Lists don't have a `.get` method, unfortunately.

sjrct replied:
:(

sjrct commented on halibot/commandmodule.py:
```diff
@@ -0,0 +1,59 @@
+	# Actually does the command handling logic. Separate callable so multi-inheritance
+	#  can work nicely (maybe)
+	def _cmd_receive(self, msg):
+		body = msg.body.split(" ",1)
```
`s/" ",1/" ", 1/`

sjrct commented on halibot/commandmodule.py:
```diff
@@ -0,0 +1,59 @@
+	#  can work nicely (maybe)
+	def _cmd_receive(self, msg):
+		body = msg.body.split(" ",1)
+		if not body[0].startswith(self.prefix):
```
Will probably keel over here if `body == ""`

richteer commented on halibot/commandmodule.py:
```diff
@@ -0,0 +1,57 @@
+	#  can work nicely (maybe)
+	def _cmd_receive(self, msg):
+		body = msg.body.split(" ", 1)
+		if body and not body[0].startswith(self.prefix):
```
One of the comments got squished in here, but hopefully that should fix the problem with empty bodies.

sjrct reviewed (approved):
I would probably favor `command-prefix` over `command_prefix` in JSON, just because I thinks it looks nicer. But overall, this seems pretty good to me.

richteer commented:
Have we decided on if we are extending jsdict-like access to config objects? Should we worry about the hyphen versus an underscore for something like `bot.config.command_prefix`?

richteer committed:
halibot: introduce CommandModule subclass of HalModule

Initial implementation of the CommandModule class of HalModule.

To implement a CommandModule:
 - Inherit from the halibot.CommandModule class
 - Define a class variable that is a
    dict: "command" -> func(self, string, msg=msg)

Example:

class Foo(CommandModule):
	def init(self):
		self.commands = {"hello", self.hello_}

	def hello_(self, string, msg=None):
		return self.reply(msg, body="Hello World!")

Note: it is recommended practice to trail command methods with an underscore
so as not to accidentally override internal halobject methods.

some fixes
